### ISSUE TITLE
Optimizing hotspot small tables

### ISSUE THREAD

**Issue by tiancaiamao** [pingcap/tidb#25293]
<table>
<tr>
    <td> <b>Focus this week</b> </td>
    <td> <b>Current status</b> </td>
</tr>
<tr>
    <td> - Disable DDL for some cases </br> - Finish performance test report</br>
 </td>
    <td> Target: Make this feature GA</br> </br>
- KR1: Finish developing  95% </br>
- KR2: Finish testing 90% </br> 
- KR3: Document, GA announce, etc 90% </td>
</tr>
<tr>
    <td> <b>Plan for the next 4 weeks </b> </td>
    <td> <b>Status indicator</b> </td>
</tr>
<tr>
    <td> Finish all testing </td>
    <td> The previous week is spend on core bank performance testing, </br>
and the result is expected </td>
</tr>
</table>

## Description

When a table is too small, it is located in just one region, that region would become a hotspot, and such hot spots would cause a performance bottleneck. By directly caching the small table data in the TiDB layer, such hot spot issues can be solved.

For a small, often used and rarely changed table, caching the whole table in memory in the TiDB server can improve performance.

## Document Collection

- Proposal doc: https://github.com/pingcap/tidb/pull/23673
- Weekly report: See the [comments of this issue](https://github.com/pingcap/tidb/issues/25293#issuecomment-950608896).

## Talent Challenge Program information

- Mentor of this issue: @tiancaiamao 
- Recommended skills:  Golang, [DDL online schema change](https://pingcap.com/blog-cn/tidb-source-code-reading-17/)(Chinese), [lease](https://www.zenlife.tk/consistent-cache.md)(Chinese) 
- Estimated Workloads: 2 Month

## Milestones and action items

Milestone 1: Support schema change for caching table, Expected finish date: TBD

- [x] #28926
- [x] #29085
    - [x] #29645
- [x] compatibility check with other feature
    - [x] #29124
    - [x] #29218
    - [x] #29318
    - [x] #29641
    - [x] #29659 
    - [x] #29837
    - [x] #29927
    - [x] #32826
- [ ] compatibility check with ecosystem tools
    - [x] #31105
    - [ ] #31190

Milestone 2: Support reading and writing on cached table, Expected finish date: TBD

- [x] #29151
- [x] #29445
    - [x] #30205
- [x] #29446
    - [x] #29546
    - [x] #29849
    - [x] #29485

Milestone 3: Test with abnormal injection cases

- [ ] Test by inject failpoint during some DDL change phase
- [ ] Chaos testing for the read/write operation

Misc:

- [x] #29958
- [x] #29959
- [x] #30950
- [x] #30962

Limitation & Known bugs

- [ ] #29682
- [ ] clean up the `mysql.table_cache_meta` table data
- [x] ~~maintain the local state to avoid unnecessary load cache request~~
- [ ] #31184

**Comment by jackwener** [comment]
Hi, I'd like to know whether it's open for working person.

**Comment by JayL-zxl** [comment]
Hi 老师您好 我申请了您负责的这个 tcp项目，我能和您email进行沟通吗

**Comment by tiancaiamao** [comment]
2021/11/01 This week:
- Two PRs for feature compatibility are merged: [admin check table](https://github.com/pingcap/tidb/pull/29096) and [stale read](https://github.com/pingcap/tidb/pull/29173)
- One PR to implement DML [read operation](https://github.com/pingcap/tidb/pull/29184) is now read for review
- And also a draft PR for [the meta lock data management](https://github.com/pingcap/tidb/pull/29152)
- One more about the feature compatibility in reivew: [forbit partition table and view](https://github.com/pingcap/tidb/pull/29256)

**Comment by tiancaiamao** [comment]
2021/11/08 This week:
- The first [PR](https://github.com/pingcap/tidb/pull/29184) for read operation is merged
- The next [PR](https://github.com/pingcap/tidb/pull/29477) to fix cache condition(issue introduced by the previous one) is merged
- Support [write operation](https://github.com/pingcap/tidb/pull/29444) is pushed and under active review, @lcwangchao gave some insightful comments,  several TODOs are found there:
    - `loadDataFromOriginalTable` should use a seperate transaction
    - The safety of "begin; sleep(1h)..." case is not well considered in the past
    - lock intend state should introduce a `oldReadLease` field to handle corner case 
    - Both read and write operation need to renew lease for their locks

**Comment by JayL-zxl** [comment]
@tiancaiamao  please add this issue into compatibility check with other feature

**Comment by tiancaiamao** [comment]
2021/11/15 This week:
- The first [PR for write operation](https://github.com/pingcap/tidb/pull/29444) is merged, it use a mocked `StateRemote` for the lock meta data management.
- Two follow up small fix for [`loadDataFromOriginalTable`](https://github.com/pingcap/tidb/pull/29575) and [`begin;sleep(1h)`](https://github.com/pingcap/tidb/pull/29443) mentioned last week is also done.
- 'alter table t nocache' is [done](https://github.com/pingcap/tidb/pull/29621), both syntax and semantic
- Support for [point get/batch point get](https://github.com/pingcap/tidb/pull/29703) is in review, and there are a couple of PRs related to compatibility is also filed and under review, like [29707](https://github.com/pingcap/tidb/pull/29707)

**Comment by tiancaiamao** [comment]
2021/11/22

* The main focus this week are following two, but both are not merged yet
    * support the renew lease operation for read [29840](https://github.com/pingcap/tidb/pull/29840)
    * use real StateRemote to replace the old mocked one [29152](https://github.com/pingcap/tidb/pull/29152)
 * The in-review PRs like point get / and add index capability are merged [29703](https://github.com/pingcap/tidb/pull/29703), [29707](https://github.com/pingcap/tidb/pull/29707)
 *  And also a new PR to disable plan cache for capability is merged in the last week [29641](https://github.com/pingcap/tidb/issues/29641)

**Comment by tiancaiamao** [comment]
2021/11/29
* The real StateRemote implementation [#29152](https://github.com/pingcap/tidb/pull/29152) finally get merged
     -- since it is too large for a single PR, it's split into two parts and the part 2 [#30066](https://github.com/pingcap/tidb/pull/30066) is still in review, 
    this has been lasting for two weeks (and a bad indicator of the project healthy)
* A PR to handle statement summary compatibility is filed this week, and in review [30166](https://github.com/pingcap/tidb/pull/30166)
* This week the SQL Infra team is in bug jail, according to our development rule @tiancaiamao can't work on the new features
    -- If the renew write lock lease is done, the mainly work should be finished, and we'll turn to testing (the next week if everything goes well)

**Comment by tiancaiamao** [comment]
2021/12/06

* No progress last week

Blocked by PR reviewing #30066, and it's been two weeks since it's in review and 3 weeks since its filed. @bb7133

**Comment by tiancaiamao** [comment]
2021/12/11
* No progress on developing
* The blocking PR [#30066](https://github.com/pingcap/tidb/pull/30066) finally get merged (after 3 weeks)
   so renew write lock lease [#30206](https://github.com/pingcap/tidb/pull/30206) is now ready for review after it.

**Comment by tiancaiamao** [comment]
2021/12/20

- [#30206](https://github.com/pingcap/tidb/pull/30206) got merged, so the development is basically done
- Planning for testing now

**Comment by tiancaiamao** [comment]
2021/12/27

- The testing plan [document](https://docs.google.com/document/d/1ux8RSo5uXd_LNOxXRATfTlNUQxSzeLwFr9t1HokV1h0/edit?usp=sharing) was written last week
- The design document [PR](https://github.com/pingcap/tidb/pull/23673) was updated and is now ready for review
- There are 2 new PRs are filed:
    - update the display of 'show create table' [#30951](https://github.com/pingcap/tidb/pull/30951)
    - introduce a `@@tidb_cache_table_lease` sys variable [#31018](https://github.com/pingcap/tidb/pull/31018)

**Comment by tiancaiamao** [comment]
2022/01/03
- This feature will not be GA in this sprint(5.4), because the testing is not finished yet
- Several PRs are filed this week for ecosystem tools compatibility [#31105](https://github.com/pingcap/tidb/pull/31105), [#31106](https://github.com/pingcap/tidb/pull/31106), [#31191](https://github.com/pingcap/tidb/pull/31191)
- 1 bug is found [#31077](https://github.com/pingcap/tidb/issues/31077) during test and fixed
- **The project will be paused for next weeks** because @tiancaiamao will be oncall in the fire brigade team

**Comment by tiancaiamao** [comment]
2022/01/17
* In the last two weeks, the project is **paused** because of the oncall. 
   This week the SQL infra team is in the bug jail again...

**Comment by tiancaiamao** [comment]
2022/02/14

There is no update for a long time during the Chinese new year (Spring festival)
Last week it was restarted.

During this period of time, several bugs are found during the performance test in the core banking scene.

- Repeated renew lease operation cause performance decrease [issue/31474](https://github.com/pingcap/tidb/issues/31474)
- Support prepared plan cache for cached table [issue/32003](https://github.com/pingcap/tidb/issues/32003)
- Aggregation push down to coprocesson conflict with cached table [issue/32157](https://github.com/pingcap/tidb/issues/32157)

They're all fixed, and a basic QPS metric for the table cache visited is added [issue/32171](https://github.com/pingcap/tidb/pull/32171)

**Comment by tiancaiamao** [comment]
2022/02/21

The performance test in the core bank scenario is done.
After caching some tables, the latency decrease by 23.8%, and throughput increase by 27.8% in RC isolation level.
And the latency decrease by 28.4%, and throughput increase by 35% in RR isolation level.

Three issues are found in the last week:
- https://github.com/pingcap/tidb/issues/32386
- https://github.com/pingcap/tidb/issues/32422
- https://github.com/pingcap/tidb/issues/32433

And I've filed PR for the first two #32387, they are under reviewing.
The last one is a performance issue, it's not cause by cached table itself, it's a long existing problem of the UnionScan executor.

**Comment by tiancaiamao** [comment]
2022/02/28

Last week there is not much progress.
3 PRs are blocked in reviewing [#32387](https://github.com/pingcap/tidb/pull/32387) [#32590](https://github.com/pingcap/tidb/pull/32590) [#32643](https://github.com/pingcap/tidb/pull/32643)
A new bug is found https://github.com/pingcap/tidb/issues/32642, it's a cornor case of the renew lease operation.
